src/core/features/tags/controller/TagsController.ts

```diff
@@ -55,24 +88,90 @@ export class TagsController {
 	}
 
 	public async add(name: string, parent: null | string): Promise<string> {
-		const db = wrapDB(this.db.get());
-
-		const segments = name.split('/');
+		validateTagName(name);
 
 		let lastId: string | null = null;
-		if (segments.length > 1) {
-			await this.db.get().transaction(async (tx) => {
-				const db = wrapDB(tx);
 
-				for (let idx = 0; idx < segments.length; idx++) {
-					const name = segments[idx];
+		await this.db.get().transaction(async (tx) => {
+			const db = wrapDB(tx);
+
+			// check tag unique
+			const {
+				rows: [duplicatedTag],
+			} = await db.query(
+				qb.line(
+					qb.sql`WITH resolved_tags AS 
+					(${qb.raw(tagsQuery, qb.sql`WHERE workspace_id = ${this.workspace}`)})
+					SELECT resolved_name FROM resolved_tags WHERE resolved_name = (${
+						parent
+							? qb.sql`(SELECT resolved_name || '/' || ${name} FROM resolved_tags WHERE id = ${parent})`
+							: qb.value(name)
+					}) 
+					LIMIT 1`,
+				),
+				z
+					.object({ resolved_name: z.string() })
+					.transform(({ resolved_name }) => ({
+						resolvedName: resolved_name,
+					})),
+			);
+			if (duplicatedTag) {
+				throw new TagControllerError(
+					`Tag ${duplicatedTag.resolvedName} already exists`,
+					TAG_ERROR_CODE.DUPLICATE,
+				);
+			}
+
+			const segments = name.split('/');
+			if (segments.length > 1) {
+				// Create an array of resolved names from a string: 'foo/bar' - ['foo', 'foo/bar']
+				const resolvedNames = segments.map((segment, index) =>
+					index === 0 ? segment : `${segments.slice(0, index + 1).join('/')}`,
+				);
+
+				// Find the existing parent tag to reuse when creating the tag hierarchy
```

Comment:
We looking for a **most qualified parent tag** here, not a parent tag.
So actually this is a root tag. Consider this in your comment and variable names